# Brooke Johnson
## Homework 1
MSDS 598 Spring 2022

In [65]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import math
import random

import seaborn as sns
import statsmodels.api as sm
from statsmodels.formula.api import ols

In [66]:
df_taxis = sns.load_dataset('taxis')
df_taxis.head()

pickup              dropoff  passengers  distance  fare   tip  \
0  2019-03-23 20:21:09  2019-03-23 20:27:24           1      1.60   7.0  2.15   
1  2019-03-04 16:11:55  2019-03-04 16:19:00           1      0.79   5.0  0.00   
2  2019-03-27 17:53:01  2019-03-27 18:00:25           1      1.37   7.5  2.36   
3  2019-03-10 01:23:59  2019-03-10 01:49:51           1      7.70  27.0  6.15   
4  2019-03-30 13:27:42  2019-03-30 13:37:14           3      2.16   9.0  1.10   

   tolls  total   color      payment            pickup_zone  \
0    0.0  12.95  yellow  credit card        Lenox Hill West   
1    0.0   9.30  yellow         cash  Upper West Side South   
2    0.0  14.16  yellow  credit card          Alphabet City   
3    0.0  36.95  yellow  credit card              Hudson Sq   
4    0.0  13.40  yellow  credit card           Midtown East   

            dropoff_zone pickup_borough dropoff_borough  
0    UN/Turtle Bay South      Manhattan       Manhattan  
1  Upper West Side South      Manhattan       Manhattan  
2           West Village      Manhattan       Manhattan  
3         Yorkville West      Manhattan       Manhattan  
4         Yorkville West      Manhattan       Manhattan

### Question 1 
Suppose we want to use the variables distance and passengers to predict total which represents the total fare. 
#### A. Use linear regression to create a model prediction total from distance and passengers. Report the ${R^2}$


In [67]:
model = ols(formula = 'total ~ distance + passengers', data=df_taxis)
res = model.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  total   R-squared:                       0.818
Model:                            OLS   Adj. R-squared:                  0.818
Method:                 Least Squares   F-statistic:                 1.450e+04
Date:                Thu, 24 Feb 2022   Prob (F-statistic):               0.00
Time:                        11:20:03   Log-Likelihood:                -20530.
No. Observations:                6433   AIC:                         4.107e+04
Df Residuals:                    6430   BIC:                         4.109e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      8.5156      0.132     64.447      0.000       8.257       8.775
distance       3.2649      0.019    170.254      0.000       3.227       3.303
passengers     0.0826      0.061      1.354      0.176      -0.037       0.202
==============================================================================
Omnibus:                     8989.441   Durbin-Watson:                   1.920
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          5652441.382
Skew:                           7.826   Prob(JB):                         0.00
Kurtosis:                     147.371   Cond. No.                         9.58
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Answer:
${R^2}$ = 0.818

#### B. Incorporate another variable taxi color into the model and then report the ${R^2}$.

In [68]:
model2 = ols(formula = 'total ~ distance + passengers + color', data=df_taxis)
res2 = model2.fit()
res2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  total   R-squared:                       0.829
Model:                            OLS   Adj. R-squared:                  0.829
Method:                 Least Squares   F-statistic:                 1.036e+04
Date:                Thu, 24 Feb 2022   Prob (F-statistic):               0.00
Time:                        11:20:03   Log-Likelihood:                -20345.
No. Observations:                6433   AIC:                         4.070e+04
Df Residuals:                    6429   BIC:                         4.073e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           5.3515      0.207     25.879      0.000       4.946       5.757
color[T.yellow]     3.8950      0.200     19.520      0.000       3.504       4.286
distance            3.2808      0.019    175.898      0.000       3.244       3.317
passengers         -0.0372      0.060     -0.625      0.532      -0.154       0.080
==============================================================================
Omnibus:                     9394.090   Durbin-Watson:                   2.034
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          6982523.917
Skew:                           8.501   Prob(JB):                         0.00
Kurtosis:                     163.502   Cond. No.                         19.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Answer:
${R^2}$ = 0.829

By incorporating taxi color into the regression, the ${R^2}$ value increased, indicating that our model does a better job at explaining the variation in total fare when we include the taxi color. Logically, this doesn't seem like it makes a lot of sense. If a researcher were approaching this problem with background research, I doubt they would theorize that the car color impacts the total fare. 

### Question 2
As in the fourth lecture, create a length of ride variable. Use the new feature you've engineered to further improve the model and report the ${R^2}$.

In [69]:
delta = pd.to_datetime(df_taxis['dropoff']) - pd.to_datetime(df_taxis['pickup'])
df_taxis['length_of_trip'] = delta / pd.Timedelta('60s')

In [70]:
model3 = ols(formula = 'total ~ distance + passengers + color + length_of_trip', data=df_taxis)
res3 = model3.fit()
res3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  total   R-squared:                       0.847
Model:                            OLS   Adj. R-squared:                  0.847
Method:                 Least Squares   F-statistic:                     8884.
Date:                Thu, 24 Feb 2022   Prob (F-statistic):               0.00
Time:                        11:20:04   Log-Likelihood:                -19985.
No. Observations:                6433   AIC:                         3.998e+04
Df Residuals:                    6428   BIC:                         4.001e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           3.3198      0.209     15.891      0.000       2.910       3.729
color[T.yellow]     3.9886      0.189     21.136      0.000       3.619       4.359
distance            2.5803      0.031     83.519      0.000       2.520       2.641
passengers         -0.0074      0.056     -0.131      0.896      -0.118       0.103
length_of_trip      0.2805      0.010     27.618      0.000       0.261       0.300
==============================================================================
Omnibus:                    10537.607   Durbin-Watson:                   2.037
Prob(Omnibus):                  0.000   Jarque-Bera (JB):         12302743.154
Skew:                          10.650   Prob(JB):                         0.00
Kurtosis:                     216.178   Cond. No.                         74.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Answer:
${R^2}$ = 0.847

By including the length of the trip we improve the explanatory power of our model but risk over-fitting by including more variables than we need.

### Question 3
Create a valifation set using ten percent of the data, and use the complement to create a training set. (Make sure there is no overlap in observation between the validation and training set).

In [83]:
from sklearn.model_selection import train_test_split

train, test = train_test_split(df_taxis, test_size=0.1)

### Question 4
Use this validation and the ${R^2}$ metric to decide whether adding length of ride to the variables from Q1 improves the model.

In [84]:
# build model on training set
model4 = ols(formula = 'total ~ distance + passengers + color + length_of_trip', data=train)
res4 = model4.fit()

# predict using the model
y_pred = res4.predict(test)

# score the model
y = test['total']
y_mean = train['total'].mean()  

TSS = sum((y_pred - y_mean)**2)
RSS = sum((y_pred - y)**2)
print(f'This model has an R^2 on the test set of {(TSS - RSS) / TSS}')

This model has an R^2 on the test set of 0.8391019564007671


#### Answer:
${R^2}$ = 0.8605

By including more variables and using validation to test our model it seems that the explanatory power is improving since the ${R^2}$ is increasing. However, we can logically surmise that some of these variables may be adding extraneous value to the ${R^2}$ metric without telling us about useful relationships, such as car color. As previously mentioned, car color logically shouldn't affect the total fare. Part of the increase in the measure of explanatory power therefore might be coming from explaining more variance that doesn't help explain the important relationships. 

### Question 5
Engineer one new feature from the data and repeat Q4. (Hint: you can use powers, differences, ratios, products and other features)

In [73]:
df_taxis.head()

pickup              dropoff  passengers  distance  fare   tip  \
0  2019-03-23 20:21:09  2019-03-23 20:27:24           1      1.60   7.0  2.15   
1  2019-03-04 16:11:55  2019-03-04 16:19:00           1      0.79   5.0  0.00   
2  2019-03-27 17:53:01  2019-03-27 18:00:25           1      1.37   7.5  2.36   
3  2019-03-10 01:23:59  2019-03-10 01:49:51           1      7.70  27.0  6.15   
4  2019-03-30 13:27:42  2019-03-30 13:37:14           3      2.16   9.0  1.10   

   tolls  total   color      payment            pickup_zone  \
0    0.0  12.95  yellow  credit card        Lenox Hill West   
1    0.0   9.30  yellow         cash  Upper West Side South   
2    0.0  14.16  yellow  credit card          Alphabet City   
3    0.0  36.95  yellow  credit card              Hudson Sq   
4    0.0  13.40  yellow  credit card           Midtown East   

            dropoff_zone pickup_borough dropoff_borough  length_of_trip  
0    UN/Turtle Bay South      Manhattan       Manhattan        6.250000  
1  Upper West Side South      Manhattan       Manhattan        7.083333  
2           West Village      Manhattan       Manhattan        7.400000  
3         Yorkville West      Manhattan       Manhattan       25.866667  
4         Yorkville West      Manhattan       Manhattan        9.533333

In [74]:
# New Feature: average speed per trip
df_taxis['trip_length_hours'] = df_taxis['length_of_trip']/60
df_taxis['average_speed'] = df_taxis['distance']/df_taxis['trip_length_hours']
df_taxis.head()

pickup              dropoff  passengers  distance  fare   tip  \
0  2019-03-23 20:21:09  2019-03-23 20:27:24           1      1.60   7.0  2.15   
1  2019-03-04 16:11:55  2019-03-04 16:19:00           1      0.79   5.0  0.00   
2  2019-03-27 17:53:01  2019-03-27 18:00:25           1      1.37   7.5  2.36   
3  2019-03-10 01:23:59  2019-03-10 01:49:51           1      7.70  27.0  6.15   
4  2019-03-30 13:27:42  2019-03-30 13:37:14           3      2.16   9.0  1.10   

   tolls  total   color      payment            pickup_zone  \
0    0.0  12.95  yellow  credit card        Lenox Hill West   
1    0.0   9.30  yellow         cash  Upper West Side South   
2    0.0  14.16  yellow  credit card          Alphabet City   
3    0.0  36.95  yellow  credit card              Hudson Sq   
4    0.0  13.40  yellow  credit card           Midtown East   

            dropoff_zone pickup_borough dropoff_borough  length_of_trip  \
0    UN/Turtle Bay South      Manhattan       Manhattan        6.250000   
1  Upper West Side South      Manhattan       Manhattan        7.083333   
2           West Village      Manhattan       Manhattan        7.400000   
3         Yorkville West      Manhattan       Manhattan       25.866667   
4         Yorkville West      Manhattan       Manhattan        9.533333   

   trip_length_hours  average_speed  
0           0.104167      15.360000  
1           0.118056       6.691765  
2           0.123333      11.108108  
3           0.431111      17.860825  
4           0.158889      13.594406

In [75]:
train, test = train_test_split(df_taxis, test_size=0.1)

# build model on training set
model5 = ols(formula = 'total ~ distance + passengers + color + average_speed', data=train)
res5 = model5.fit()

# predict using the model
y_pred = res5.predict(test)

# score the model
y = test['total']
y_mean = train['total'].mean()  

TSS = sum((y_pred - y_mean)**2)
RSS = sum((y_pred - y)**2)
print(f'This model has an R^2 on the test set of {(TSS - RSS) / TSS}')

This model has an R^2 on the test set of nan


#### Answer:
${R^2}$ = 0.911

The ${R^2}$ improved quite a lot, and the model looks like it has very high explanatory power from adding in the average speed of each trip. 

### Question  6
Why shouldn't you use fare or tip to predict total?

#### Answer:
Because the fare and tip are the parts that compose the total. 

$total = fare*distance + tip$ where $tip = percent * (fare*distance)$

Therefore the two variables, fare and tip, have high multicolinearty which can produce less reliable results by increasing the confidence interval. 
